tb_plugin/torch_tb_profiler/profiler/overall_parser.py

```diff
@@ -274,6 +332,139 @@ def update_steps_consider_device_side(self, runtime_node_list, device_node_list)
+                    elif current_bucket[0] >= current_range[1]:
+                        current_range_index += 1
+                        if current_range_index < len(self.kernel_ranges_per_device[i]):
+                            current_range = self.kernel_ranges_per_device[i][current_range_index]
+                    else:
+                        left_bound = max(current_range[0], current_bucket[0])
+                        right_bound = min(current_range[1], current_bucket[1])
+                        gpu_utilization_timeline[-1][current_bucket_index] += (right_bound - left_bound)
+                        if current_bucket[1] < current_range[1]:
+                            current_bucket_index += 1
+                            current_bucket = (all_steps_range[0] + current_bucket_index * bucket_size,
+                                              all_steps_range[0] + (current_bucket_index + 1) * bucket_size)
+                        else:
+                            current_range_index += 1
+                            if current_range_index < len(self.kernel_ranges_per_device[i]):
+                                current_range = self.kernel_ranges_per_device[i][current_range_index]
+                for i_bucket in range(buckets):
+                    gpu_utilization_timeline[-1][i_bucket] /= bucket_size
+
+                for i_bucket in range(buckets):
+                    start_time = self.steps[0][0] + i_bucket * bucket_size
+                    util_json = build_trace_counter(device_id, start_time, gpu_utilization_timeline[-1][i_bucket])
+                    counter_json.append(util_json)
+                start_time = self.steps[0][0] + buckets * bucket_size
+                util_json = build_trace_counter(device_id, start_time, 0)
+                counter_json.append(util_json)
+
+        self.kernel_ranges_per_device = None  # Release memory.
+
+        return counter_json
+
+
+    def calculate_approximated_sm_efficency(self):
+        def calculate_avg(approximated_sm_efficency_ranges, total_dur):
+            total_weighted_sm_efficiency = 0.0
+            for r in approximated_sm_efficency_ranges:
+                dur = r[0][1] - r[0][0]
+                total_weighted_sm_efficiency += r[1] * dur
+            avg_approximated_sm_efficency = total_weighted_sm_efficiency / total_dur
+            return avg_approximated_sm_efficency
+
+        def build_trace_counter(gpu_id, start_time, counter_value):
+            util_json = {}
+            util_json["ph"] = "C"
+            util_json["name"] = "GPU {} Est. SM Efficiency".format(gpu_id)
+            util_json["pid"] = 0
+            util_json["ts"] = start_time
+            util_json["args"] = {"Est. SM Efficiency": counter_value}
+            return util_json
+
+        counter_json = []
+        total_dur = self.steps[-1][1] - self.steps[0][0]
+        approximated_sm_efficency_ranges_per_device = []
+        for device_id, i in self.device_to_index.items():
```

Comment:
if the device_ids is list, then for device_id, in in enumerate(self.device_ids): should be enough?

Reply:
Thanks. Now change to a set.